File: 2dsearch.py
# // Time Complexity :O(m+n),m is the row n is the column
# // Space Complexity :O(1)
# // Did this code successfully run on Leetcode :yes
# // Any problem you faced while coding this :no


# // Your code here along with comments explaining your approach




class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        
        i=0
        j=len(matrix[0])-1
        while(i<len(matrix) and j>=0):
                
                if matrix[i][j]==target:
                    return True
                    
                if matrix[i][j]>target:
                    j=j-1
                else:
                    i=i+1
        return False

File: 2DSearch.java
//Time Complexity: O(log(m+n))
//space complexity: O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english

// reduce this problem to binary search
// consider first row and last col as one array, we can observe that all items are sorted
// so compare with mid and shift left or down  

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length;
        int n= matrix[0].length;
        
        //represent row number as r and col number as c
        
        //initially we start with first row last item (0,n-1)
        int r=0,c=n-1;
        
        //while r and c are within limits 
        while(r<m && c>=0) {
            
            //our element will always be at the intersection of r and c
            int mid = matrix[r][c];
            
            if (mid == target) {
                return true;
            } 
            
            if(target < mid) { //in case target less than mid
                c--; //we move left --> that is prev col
            } else {
                r++; // we move down --> to next row
            }
            
        }
        
        //we have not found the element
        return false;
    }
}